### PR TITLE
[receiver/skywalking]  Create  skywalking component folder/structure

### PR DESCRIPTION
**Pull request by JaredTan95**
**Description:** <Describe what has changed.>

1st PR: Create  skywalking component folder/structure and finish factory , config unit test.

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/7954
**Testing:** <Describe what testing was performed and which tests were added.>

**Documentation:** <Describe the documentation added.>
added skywalking receiver `README.md`

### CODE REVIEW

File: receiver/skywalkingreceiver/config.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package skywalkingreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/skywalkingreceiver"
+
+import (
+	"fmt"
+	"go.opentelemetry.io/collector/config"
```
Comment (pmm-sumo) [inline receiver/skywalkingreceiver/config.go:19]: Import order needs to be fixed here

File: receiver/skywalkingreceiver/config_test.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
+package skywalkingreceiver
+
+import (
+	"github.com/stretchr/testify/assert"
```
Comment (pmm-sumo) [inline receiver/skywalkingreceiver/config_test.go:4]: Imports order needs to be fixed here as well. `make goimports` should help with it

Comment (JaredTan95) [reply receiver/skywalkingreceiver/config_test.go:4]: ok, thanks

Comment (pmm-sumo) [reply receiver/skywalkingreceiver/config_test.go:4]: You can run `make gofmt`, which should take care of that :)

File: receiver/skywalkingreceiver/factory.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+	// Convert settings in the source config to configuration struct
+	// that Skywalking receiver understands.
+	rCfg := cfg.(*Config)
+
+	var config configuration
+	// Set ports
+	if rCfg.Protocols.GRPC != nil {
+		config.CollectorGRPCServerSettings = *rCfg.Protocols.GRPC
+		config.CollectorGRPCPort, _ = extractPortFromEndpoint(rCfg.Protocols.GRPC.NetAddr.Endpoint)
```
Comment (pmm-sumo) [inline receiver/skywalkingreceiver/factory.go:87]: Error is not checked. Should the function fail when there is error returned?

File: receiver/skywalkingreceiver/factory.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	if rCfg.Protocols.HTTP != nil {
+		config.CollectorHTTPSettings = *rCfg.Protocols.HTTP
+		config.CollectorHTTPPort, _ = extractPortFromEndpoint(rCfg.Protocols.HTTP.Endpoint)
```
Comment (pmm-sumo) [inline receiver/skywalkingreceiver/factory.go:92]: Similarly as above, error is not checked here. Should it return `nil, err` in case `extractPortFromEndpoint` fails?

File: receiver/skywalkingreceiver/config.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+var _ config.Receiver = (*Config)(nil)
+var _ config.Unmarshallable = (*Config)(nil)
+
+// Validate checks the receiver configuration is valid
+func (cfg *Config) Validate() error {
+	if cfg.GRPC == nil && cfg.HTTP == nil {
+		return fmt.Errorf("must specify at least one protocol when using the Skywalking receiver")
+	}
+
```
Comment (pmm-sumo) [inline receiver/skywalkingreceiver/config.go:50]: I think it would be good to validate parsing the grpc or http configuration as well

File: receiver/skywalkingreceiver/factory.go (hunk 3 above)
Comment (JaredTan95) [reply receiver/skywalkingreceiver/factory.go]: I added more err checks.

File: receiver/skywalkingreceiver/config.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+		return err
+	}
+
+	protocols, err := componentParser.Sub(protocolsFieldName)
+	if err != nil {
+		return err
+	}
+
+	if !protocols.IsSet(protoGRPC) {
+		cfg.GRPC = nil
```
Comment (pmm-sumo) [inline receiver/skywalkingreceiver/config.go:86]: Not sure, why is this required?

File: receiver/skywalkingreceiver/tracing_report_service.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+)
+
+const (
+	httpEventName = "http-tracing-event"
+	grpcEventName = "grpc-tracing-event"
+	success       = "success"
+	failing       = "failing"
+)
+
+type TraceSegmentReportService struct {
```
Comment (pmm-sumo) [inline receiver/skywalkingreceiver/tracing_report_service.go:33]: Does this struct (as well as the functions below) need to be public?

File: receiver/skywalkingreceiver/factory.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+		}
+	}
+
+	// Create the receiver.
+	return newSkywalkingReceiver(rCfg.ID(), &c, nextConsumer, set), nil
+}
+
+// extract the port number from string in "address:port" format. If the
+// port number cannot be extracted returns an error.
+func extractPortFromEndpoint(endpoint string) (int, error) {
```
Comment (pmm-sumo) [inline receiver/skywalkingreceiver/factory.go:106]: This is probably not a topic for this PR, but I noticed this function is copied across several components. Perhaps it should be extracted into some common module?

File: receiver/skywalkingreceiver/config.go (hunk 6 above)
Comment (JaredTan95) [reply receiver/skywalkingreceiver/config.go:86]: just a redundant check.

File: receiver/skywalkingreceiver/tracing_report_service.go (hunk 7 above)
Comment (JaredTan95) [reply receiver/skywalkingreceiver/tracing_report_service.go:33]: Yes, it should be private. I will update it.

File: receiver/skywalkingreceiver/factory.go (hunk 8 above)
Comment (JaredTan95) [reply receiver/skywalkingreceiver/factory.go:106]: make sense, but I am a beginner for `otel collector `, Would you tell me where or which module it should be？

Comment (JaredTan95) [reply receiver/skywalkingreceiver/factory.go:106]: I guess may be `internal/tools/` or `internal/common`?

Comment (pmm-sumo) [reply receiver/skywalkingreceiver/factory.go:106]: This is more like a general observation, I am wondering what is @jpkrohling opinion on that :)

File: receiver/skywalkingreceiver/config.go (hunk 6 above)
Comment (pmm-sumo) [reply receiver/skywalkingreceiver/config.go:86]: I think it's more than just a check. The way it works is that if `grpc` is not included on the protocol list, the receiver will disable it.

This is not consistent with how `http` is used. It will be always started (if not provided specific config, then using the defaults). 

If this was intended to map how e.g. Jaeger receiver works, I think similar check and setting nil is required for HTTP as well.

File: receiver/skywalkingreceiver/README.md [hunk 9]
```diff
@@ -0,0 +1,24 @@
+# Skywalking Receiver
+
+Receives trace data in [Skywalking](https://skywalking.apache.org/) format.
+
+Supported pipeline types: traces
+
+## Getting Started
```
Comment (pmm-sumo) [inline receiver/skywalkingreceiver/README.md:7]: I'm missing a section with the list of protocols, their defaults and which one is served automatically or not, e.g. [like for Jaeger](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/jaegerreceiver#getting-started)

File: receiver/skywalkingreceiver/README.md [hunk 10]
```diff
@@ -0,0 +1,24 @@
+# Skywalking Receiver
+
+Receives trace data in [Skywalking](https://skywalking.apache.org/) format.
+
+Supported pipeline types: traces
```
Comment (pmm-sumo) [inline receiver/skywalkingreceiver/README.md:5]: We might want to add the status of the receiver or at least mention it's not yet fully functional

File: receiver/skywalkingreceiver/config.go (hunk 6 above)
Comment (JaredTan95) [reply receiver/skywalkingreceiver/config.go:86]: yes, I referred to the implementation of jaeger receiver. I will update it.

File: receiver/skywalkingreceiver/testdata/config.yaml [hunk 11]
```diff
@@ -0,0 +1,31 @@
+receivers:
+#  otlp:
```
Comment (pmm-sumo) [inline receiver/skywalkingreceiver/testdata/config.yaml:2]: Is this commented out receiver intended? I think it can be simply removed

File: receiver/skywalkingreceiver/config.go [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+		}
+	}
+
+	return nil
+}
+
+// Unmarshal a config.Parser into the config struct.
+func (cfg *Config) Unmarshal(componentParser *config.Map) error {
+	if componentParser == nil || len(componentParser.AllKeys()) == 0 {
+		return fmt.Errorf("empty config for Jaeger receiver")
```
Comment (pmm-sumo) [inline receiver/skywalkingreceiver/config.go:70]:
```suggestion
		return fmt.Errorf("empty config for Skywalking receiver")
```

File: receiver/skywalkingreceiver/factory_test.go [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+	factory := NewFactory()
+	cfg := factory.CreateDefaultConfig()
+	assert.NotNil(t, cfg, "failed to create default config")
+	assert.NoError(t, configtest.CheckConfigStruct(cfg))
+}
+
+func TestCreateReceiver(t *testing.T) {
+	factory := NewFactory()
+	cfg := factory.CreateDefaultConfig()
+	// have to enable at least one protocol for the jaeger receiver to be created
```
Comment (pmm-sumo) [inline receiver/skywalkingreceiver/factory_test.go:51]:
```suggestion
	// have to enable at least one protocol for the skywalking receiver to be created
```

File: receiver/skywalkingreceiver/trace_receiver.go [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+// configuration defines the behavior and the ports that
+// the Skywalking receiver will use.
+type configuration struct {
+	CollectorHTTPPort           int
+	CollectorHTTPSettings       confighttp.HTTPServerSettings
+	CollectorGRPCPort           int
+	CollectorGRPCServerSettings configgrpc.GRPCServerSettings
+}
+
+// Receiver type is used to receive spans that were originally intended to be sent to Jaeger.
```
Comment (pmm-sumo) [inline receiver/skywalkingreceiver/trace_receiver.go:49]:
```suggestion
// Receiver type is used to receive spans that were originally intended to be sent to Skywaking.
```

File: receiver/skywalkingreceiver/trace_receiver.go [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+// This receiver is basically a Jaeger collector.
```
Comment (pmm-sumo) [inline receiver/skywalkingreceiver/trace_receiver.go:50]:
```suggestion
// This receiver is basically a Skywalking collector.
```

File: receiver/skywalkingreceiver/trace_receiver.go [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+			if errHTTP := sr.collectorServer.Serve(cln); !errors.Is(errHTTP, http.ErrServerClosed) && errHTTP != nil {
+				host.ReportFatalError(errHTTP)
+			}
+		}()
+	}
+
+	if sr.collectorGRPCEnabled() {
+		opts, err := sr.config.CollectorGRPCServerSettings.ToServerOption(host, sr.settings.TelemetrySettings)
+		if err != nil {
+			return fmt.Errorf("failed to build the options for the Jaeger gRPC Collector: %v", err)
```
Comment (pmm-sumo) [inline receiver/skywalkingreceiver/trace_receiver.go:168]:
```suggestion
			return fmt.Errorf("failed to build the options for the Skywalking gRPC Collector: %v", err)
```

File: receiver/skywalkingreceiver/testdata/config.yaml (hunk 11 above)
Comment (JaredTan95) [reply receiver/skywalkingreceiver/testdata/config.yaml]: ok

**Comment by JaredTan95** [comment]
thanks @pmm-sumo  again

**Review by pmm-sumo — APPROVED** [review]
I think it looks good for the first phase now. Lets see what @jpkrohling and others say :)

File: receiver/skywalkingreceiver/trace_receiver.go [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+func (sr *swReceiver) collectorHTTPEnabled() bool {
+	return sr.config != nil && sr.config.CollectorHTTPPort > 0
+}
+
+func (sr *swReceiver) Start(_ context.Context, host component.Host) error {
+
+	return sr.startCollector(host)
+}
+
+func (sr *swReceiver) Shutdown(ctx context.Context) error {
```
Comment (jpkrohling) [inline receiver/skywalkingreceiver/trace_receiver.go:121]: Don't you have a test for the Start/Shutdown routine?

**Comment by JaredTan95** [comment]
> This is also failing:
> 
> ```
> Using versioning file /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/versions.yaml
> verifyAllModulesInSet failed: Module github.com/open-telemetry/opentelemetry-collector-contrib/receiver/skywalkingreceiver (defined in /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/skywalkingreceiver/go.mod) is not listed in any module set.
> ```
> 
> Please, go over the contribution guidelines and check which files need to be changed for new components.

I did not change it , especially.  because this is the 1st PR and the receiver does not fully functional.
I will update this, thanks.
